### ISSUE TITLE
upgrade command should work on Windows

### ISSUE THREAD

**Issue by twpayne** [twpayne/chezmoi#1605]
## Is your feature request related to a problem? Please describe.

Currently, `chezmoi upgrade` only works on a few operating systems, and notably it is [completely disabled on Windows](https://github.com/twpayne/chezmoi/blob/83bad0ac556fae34688e9db3821f55f031745d36/internal/cmd/noupgradecmd.go#L1).

## Describe the solution you'd like

`chezmoi upgrade` should work on Windows, just as it does on UNIX-like operating systems.

## Describe alternatives you've considered

Leaving things as they are.

## Additional context

@felipecrs and I discussed this earlier today. Right now, there are a few complications for making `chezmoi upgrade` work on Windows:

1. As I currently understand it (and I might be wrong), Windows does not allow you to replace a file while it is opened, which means that a chezmoi binary cannot replace it itself while it is running without some extra external mechanism.
2. Detecting exactly how chezmoi was installed on Windows is not obvious. For example, chezmoi might be installed via @zb140's [PowerShell script](https://github.com/twpayne/chezmoi/blob/master/assets/scripts/install.ps1), via [Chocolatey](https://github.com/twpayne/chezmoi/pull/759), via [Scoop](https://github.com/twpayne/scoop-bucket/blob/master/chezmoi.json), or via some other method.

@felipecrs points out that the `deno upgrade` command works around problem 1 above using [this logic](https://github.com/denoland/deno/blob/22dd1b90c4eb5671decd458cc70935898f9ccf0d/cli/tools/upgrade.rs#L286-L298). Maybe chezmoi can use the same?

cc @zb140 as the Windows expert :)

**Comment by felipecrs** [comment]
Another option would be to delegate the replacing part to a PowerShell or CMD process, so that chezmoi.exe would not be in use during the replacing. 

It's possible to do such kind of thing (replace current process with another process, thus closing the current process) in Linux in go with [`syscall.Exec`](https://stackoverflow.com/a/17356031/12156188), but I don't know how it would be done on Windows.

**Comment by zb140** [comment]
Yeah, it would be great to make this work.  When I first started filling out the Windows support I intended to do it, but by the time I got most of the rest of the features working it had sort of fallen by the wayside :smile:

Deno's solution seems like a reasonable one.  I'm actually mildly surprised that Windows lets you do that, but the important part is that it works, and it would be a totally reasonable way to implement upgrade in a lot of cases, including the PowerShell script install.  For Scoop, a reasonable heuristic might be: if `scoop` is in `$PATH` and `scoop which`  returns something that "matches" what `os.Executable()` says (for some appropriate definition of "match") just run `scoop update chezmoi`.  I don't know for sure if scoop will be able to do the update while `chezmoi.exe` is running, but I _think_ it will work because of the way Scoop's shims work.

Chocolatey would be a bit more complicated, depending on exactly how much detail we wanted to go into.  There isn't a direct equivalent of the `which` command yet (but there's been some discussion of adding something similar, see https://github.com/chocolatey/choco/issues/681). In the meantime, there's an XML file in `C:\ProgramData\chocolatey\.chocolatey\chezmoi.[version]\.files` that could potentially be parsed to find the location of `chezmoi.exe`.  That path should be (fairly) stable but could itself could maybe be parsed from the output of `choco list --local-only chezmoi --trace` ... I think that's probably getting a little too complicated.

**Comment by felipecrs** [comment]
For Chocolatey, I suggest:

1. Check if `ChocolateyInstall` environment variable exists. If not, it's not installed through Chocolatey
2. If yes, appends `\bin\chezmoi.exe` to it and compare it against the chezmoi bin's PATH
3. If matches, `choco upgrade chezmoi` (as Administrator :D)

**Comment by zb140** [comment]
That sounds good.  I don't use Chocolatey much anymore, so I forgot about the "as Administrator" part, but it's a good call-out :smile: Scoop doesn't require admin privileges, and I imagine most instances where we'd just download the new binary shouldn't either (unless someone has set things up very weird...)

**Comment by felipecrs** [comment]
> (unless someone has set things up very weird...)

Yeah... in case someone places the binary in a system folder like Program Files.

For reference: https://stackoverflow.com/a/26068273/12156188

But if we end up by using the PowerShell delegation approach, then `PowerShell.exe -RunAs` would also work.

**Comment by bradenhilton** [comment]
Wouldn't it make more sense to only have `upgrade` work if `chezmoi` was installed via the install script? If I installed a package via a package manager, I would expect to also upgrade that package using the package manager. Upgrading a package outside of a package manger can also break things depending on well the package manager can handle this.

If you don't want to have separate builds for Windows (upgradeable and non-upgradeable) perhaps you could add an option in the config to disable self-upgrades?

For example:

```toml
[upgrade]
    enabled = 'no' # off/false/0 etc.
```

**Comment by zb140** [comment]
In the package manager case, it would still be the package manager doing the actual upgrade; we'd just try to detect which one was originally used and execute the appropriate command to perform the upgrade. There may be an edge case or two with Chocolatey since there doesn't seem to be an official way to list the files installed by a given package, but it seems unlikely to be a common issue -- in order to cause a problem, the detection would have to conclude that `chezmoi.exe` was _not_ installed by Chocolatey when it really was.  For _that_ to happen, you'd have to arrange for Chocolatey to install somewhere outside of `$env:ChocolateyInstall\bin`. I'm admittedly not the most familiar with Chocolatey, but I have no idea how one might actually accomplish that.  In any case, I think it's reasonable to say that if you're in that situation, it's because you tried pretty hard to get yourself into it.  If it ever became clear that this was causing a real-world issue for some `chezmoi` users, we could always reconsider implementing the more complicated logic using the XML file I mentioned above. Since it lists the full path of the installed files, it wouldn't have this particular problem.

**Comment by bradenhilton** [comment]
@zb140 You can force the issue by installing `chezmoi` via Chocolatey then simply uninstalling Chocolatey. Doing so leaves behind the `ChocolatelyInstall` environment variable and the `chezmoi` binary, which means @felipecrs's steps above are not reliable enough in this case. You could argue that doing that would be silly, and I'd agree with you, but it would still result in a fully functional installation of `chezmoi` which now cannot be upgraded unless the user reinstalls Chocolatey. I'd assume you can do something similar with Scoop as well.

More concerns/edge cases:

- What should happen if multiple different installations of `chezmoi` are found on the machine? Perhaps with differing installation methods? Should they all be upgraded? Should `upgrade` only upgrade the first result and leave the rest alone?

- What if a package manager changes how it upgrades packages in a new major version? Should `chezmoi` start checking for the versions of each package manager to know how to properly invoke an upgrade?

Why not just detect the installation method and inform the user of how to upgrade `chezmoi` via the package manager if that's what was detected? You avoid basically all edge cases this way, or at least the particularly nasty ones.

The only installations we have direct "control" over are the static binaries installed either manually or via the install script. I don't believe it's `chezmoi`'s place to start invoking package managers on behalf of the user, though of course we can agree to disagree.

**Comment by felipecrs** [comment]
> which means @felipecrs's steps above are not reliable enough in this case.

That can be simply fixed by adding an additional step: looking for `choco.exe` in PATH.

However, in such situation that chezmoi would recognize that it was installed by Chocolatey (and it really was), at most it would fail when running `choco upgrade chezmoi` with `choco not found` or something like that, which I guess is a completely okay error message to the user, and even expected (because the user was the one who uninstalled Chocolatey with chezmoi left to begin with).

> What should happen if multiple different installations of chezmoi are found on the machine?

Only the current chezmoi in execution should be taken care of. My method says to compare the executable path of the current chezmoi only, not to check for any other chezmoi entries in the PATH.

> Should they all be upgraded?

Of course no, I guess. And why would a user have such situation to begin with? Remember the `chezmoi upgrade` is just provided for convenience, it should not come with the compromise of handling every single corner case.

> What if a package manager changes how it upgrades packages in a new major version? Should chezmoi start checking for the versions of each package manager to know how to properly invoke an upgrade?

This is very unlikely, Chocolatey did it (previously it was `update` and now it's `upgrade`), but even so `update` still works. And yes, if that problem ever happens to happen, checking the version of the package manager before issuing the command should be a good approach.

> Why not just detect the installation method and inform the user of how to upgrade chezmoi via the package manager if that's what was detected?

If chezmoi would already know how to upgrade to inform the user, why don`t do it by itself? If the problem is the concerns about the corner cases, the following message should be good enough:

```console
$ chezmoi upgrade
chezmoi detected to be installed with Chocolatey, and it will now run the following command on your behalf to upgrade it.

  choco upgrade chezmoi --yes

You can also run the above command by yourself if you prefer. Confirm?
```

> The only installations we have direct "control" over are the static binaries installed either manually or via the install script. I don't believe it's chezmoi's place to start invoking package managers on behalf of the user, though of course we can agree to disagree.

Well, I really know what you mean and I kind of agree. But perhaps chezmoi can be an exception of the rule, because chezmoi itself is oftenly used to manage the package managers as well (like having a run_once script which runs `apt install zsh`) or something like that.

I'm especially worried about: https://github.com/twpayne/chezmoi/issues/1600

But honestly, your suggestion is an interesting compromise to take:

1. `chezmoi upgrade` to works only if the binary was installed without a package manager. An output like `chezmoi was intalled by other means, please use the upgrade method of the package manager which you used to install it to upgrade chezmoi`.

And then, when auto "self-upgrading" (the feature request above), it can output a similar error message: `a newer version of chezmoi is required, and chezmoi cannot self-upgrade because it was installed by a package manager`

**Comment by zb140** [comment]
Hm, that's a good call-out. I'm a _little_ surprised that uninstalling Chocolatey doesn't remove the environment variable, but it does make some sense.  Also, in that situation, chezmoi wouldn't be easily uninstallable either, since there's no real installer and it wouldn't show up in Add/Remove Programs.

For multiple installations, I would expect that `chezmoi upgrade` would only ever try to upgrade the version that is actually executing.  Rather than searching for installed instances of `chezmoi.exe`, the process would look more like "was the currently running instance installed by Scoop? If not, was it installed by Chocolatey?" etc.  If you wanted to upgrade any other instances, you'd have to do it manually using whatever process you used to install in the first place.

New package manager versions is another potential concern, but at least in my mind, `chezmoi upgrade` is a convenience rather than the only (possibly not even the _best_) method of upgrading.  So to me, if we determine we don't know how to upgrade, or if the attempt to upgrade fails, we can always fall back to informing the user as you suggest.  In any case, I don't feel super strongly about whether or not package manager upgrades should be supported or not -- I know I installed chezmoi with Scoop, and `scoop upgrade chezmoi` is just as easy :smile:  But I do think it would be nice to have the consistency with the way `chezmoi upgrade` works on other platforms.

**Comment by twpayne** [comment]
`chezmoi upgrade` is best-effort and there's no way that it will be reliable in all circumstances. It's primarily a helper for people who have used the one-line shell install script or downloaded and installed a package from the releases page. I assume that people who are already using a package manager will use that same package manager to upgrade chezmoi and will never use the `chezmoi upgrade` command.

For packaged versions of chezmoi, package maintainers can build chezmoi with the `noupgrade` build tag which disables the `chezmoi upgrade` command completely, effectively forcing users to use the package manager to upgrade chemzoi.

**Comment by twpayne** [comment]
Fixed by #1628.
